lib/ABI/FunctionType/Conversion.cpp

```diff
@@ -327,11 +327,12 @@ static bool verifyAlignment(const abi::Definition &ABI,
   }
 }
 
-static bool canBeNext(ArgumentDistributor &Distributor,
-                      model::QualifiedType CurrentType,
-                      uint64_t CurrentOffset,
-                      uint64_t NextOffset,
-                      uint64_t NextAlignment) {
+template<typename TypeOfTheType>
```

Comment:
Put a constraint on this to be either a `QualifiedType` or a `Type *`?

Reply:
```cpp
template<typename T>
concept ModelTypeLike = requires(const T &V, const abi::Definition &ABI) {
  { V.size() } -> std::convertible_to<std::optional<uint64_t>>;
  { ABI.alignment(V) } -> std::convertible_to<std::optional<uint64_t>>;
  { ABI.hasNaturalAlignment(V) } -> std::convertible_to<std::optional<bool>>;
};
```
?

Reply:
I'd just do `derives_from<Type *> or same_as<QualifiedType>`.
What you propose would make sense if we ever expect anything more than `Type *` and `QualifiedType`, but I don't think it's going to be the case.

Reply:
> but I don't think it's going to be the case.

That's why I asked instead of just adding it. We'll eventually reshuffle type-types, won't we? This way it only depends on the specific functionality these functions need as opposed to type names - it's a lot harder for this to get desyncronized.

Reply:
I think that in the effort of inlining primitive types, having `derives_from<Type *> or same_as<QualifiedType>` will make sure we stop and looking into that.

Asserting the compatibility with another API is a bit more cumbersome to me. Let's keep it simple.

Reply:
Done.